Theme Showcase Search Revamp

blackjackkent opened the pull request:
#### Changes proposed in this Pull Request

* Rework the theme showcase search
  * All / Free / Premium buttons - reworked
  * Advanced search toggle - reworked
  * Clear button - now inside
  * Suggestions - now floating

#### BEFORE

![2021-07-14_13-30](https://user-images.githubusercontent.com/937354/125674048-32e33cc1-048c-4e33-b13c-69e98a85ac0f.png)
^ Before - Initial

![2021-07-14_13-30_1](https://user-images.githubusercontent.com/937354/125674083-300f82b5-2845-4dc2-8666-5059a177a7d7.png)
^ Before - Magic Welcome

![2021-07-14_13-30_2](https://user-images.githubusercontent.com/937354/125674108-dd885984-c2da-4393-bf55-1e5cb016669a.png)
^ Before - Keyed Suggestions

#### AFTER

![2021-07-15_10-03](https://user-images.githubusercontent.com/937354/125810925-840d313e-cf7f-4b44-ad6b-5c4e8e37431d.png)
^ After - Initial

![2021-07-15_10-04](https://user-images.githubusercontent.com/937354/125810934-65f988d8-203c-48a7-a78b-edde27e61697.png)
^ After - Magic Welcome (Now in popover)

![2021-07-15_10-04_1](https://user-images.githubusercontent.com/937354/125810954-2c57d5af-aa65-4911-93b2-c6ea689872af.png)
^ After - Keyed Suggestions (Now floating)

#### Testing instructions

* Go to theme showcase and use the search
  * See suggestions while typing in 'sidebar'
  * All/Free/Premium filters
  * Filter button -> Features/Column/Subject

Related to #54392

blackjackkent committed:
Rearranging magic search bar for cleaner appearance

mreishus commented:
Make KeyedSuggestions float #54515 - Merges into this PR.

mreishus committed:
Make the KeyedSuggestions Floating (#54515)

mreishus committed:
Magic welcome bar is moved to popover (#54525)

* Make the KeyedSuggestions Floating

* Magic search welcome bar moved to popover

mreishus commented:
Added draft version of floating suggestions, and popover for magic search bar (needs design tweaks)
![2021-07-13_14-03](https://user-images.githubusercontent.com/937354/125510420-08971134-ff54-412a-8206-00e45f3e1d64.png)

mreishus committed:
Clean up isWelcomeBarEnabled

blackjackkent committed:
Updating mobile styling for magic search bar

mreishus commented:
Current issues:
* Right side of the floating suggestions menu is 1px off alignment
* ~~Long text fade is broken when certain elements are focused~~
  * Caused by css conflict after moving keyedsuggestions in dom. Fixed here: `[update/rearrange-magic-search-bar` 6f8ab1d48b`] keyedSuggestions: has-highlight -> is-selected`

![Peek 2021-07-13 14-27-long-text-bug](https://user-images.githubusercontent.com/937354/125513225-a24c14e1-fa6c-4475-af8d-12359b840276.gif)

* ~~Content of filters popover doesn't match mockup~~ (Edit: They now match somewhat, but not exactly)

![2021-07-13_10-04](https://user-images.githubusercontent.com/937354/125513385-f3c03c8e-71e7-4ce4-9b6f-1c9e2dbcfe6b.png)
^ Mock up
![2021-07-13_14-28](https://user-images.githubusercontent.com/937354/125513353-1af58660-9bb8-42fb-9da2-8cc44f0daee8.png)
^ State of this branch ~~now~~ earlier
![2021-07-13_16-53](https://user-images.githubusercontent.com/937354/125530564-048ae5fb-8989-43d3-862a-93bbec959022.png)
^ State of this branch now


* ~~Filters button isn't 'inside' search bar'~~
  * Fixed here: `[update/rearrange-magic-search-bar `854fe85556`] Move filter button inside search`

* Extra border below search
  * These are box shadows and seem to be difficult to remove only the bottom side on
  * ![2021-07-13_15-37](https://user-images.githubusercontent.com/937354/125521484-d6a6a11e-3abd-47bf-a388-9716f034bbff.png)

* ~~"Clear" button is awkward~~
![2021-07-13_14-32](https://user-images.githubusercontent.com/937354/125513764-9b1b17ff-7e0d-4303-9b90-aa0e65d3953d.png)
  * Moved inside search bar here `[update/rearrange-magic-search-bar `b4ffc69ec9`] Move clear button inside search`
  * ![2021-07-13_15-36](https://user-images.githubusercontent.com/937354/125521304-e9f60bbf-c57f-49c5-a54a-39ee71d505a9.png)

* Keyed Suggestions doesn't overlap the banner completely, even if I max out the z-index
  * ![2021-07-13_16-54](https://user-images.githubusercontent.com/937354/125530438-898a2794-f7c7-45e7-8634-d57088c0b66a.png)

mreishus committed:
keyedSuggestions: has-highlight -> is-selected

"has-highlight" is already a css class used by the search bar, which was
conflicting and causing a bug

mreishus committed:
Add closePopover callback

sixhours commented on client/my-sites/themes/themes-magic-search-card/index.jsx:
```diff
@@ -290,15 +294,17 @@ class ThemesMagicSearchCard extends React.Component {
-				placeholder={ translate(
-					'Search by style or feature: portfolio, store, multiple menus, or…'
-				) }
+				placeholder={ translate( 'Search for themes' ) }
```
Hmm, we just updated this wording based on HE feedback in #54170, but the mockup uses "Search for themes". Is the wording update intentional? /cc @onuro

mreishus replied:
I'm assuming this is not intentional, so I've reverted it for now: 
`[update/rearrange-magic-search-bar `09a077b5c1`] Restore original <Theme placeholder text`

mreishus committed:
Move clear button inside search

mreishus committed:
Move filter button inside search

mreishus committed:
Style changes to magic search welcome

mreishus committed:
Restore original <Theme placeholder text

mreishus commented:
Current issues:

- Right side of the floating suggestions menu is 1px off alignment
- ~~Extra border below search~~ Fixed
  - These are box shadows and seem to be difficult to remove only the bottom side on
- ~~Keyed Suggestions doesn't overlap the banner completely, even if I max out the z-index~~ Fixed: 7393c63
  - Some portions of the upwork banner have a (lowish) z-index set, like 2 or 4. See code: https://github.com/Automattic/wp-calypso/pull/31195/files
    - We should probably use the same mechanism for setting zindex via scss `assets/stylesheets/shared/functions/_z-index.scss` in this PR too
  - This is failing because `.themes-magic-search-card.has-highlight` has a z-index of 1: [Link](https://github.com/Automattic/wp-calypso/blob/09a077b5c1695cc146caccfae82a49b025b370ce/client/my-sites/themes/themes-magic-search-card/style.scss#L25).
    - This modifies the [Stacking context](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Positioning/Understanding_z_index/The_stacking_context) of all child elements, including keyed suggestions, which means that setting the z-index to 99999999 is still constrained by the parent which has a z index of 1.

mreishus committed:
Fix z-index issues

sixhours commented:
> These are box shadows and seem to be difficult to remove only the bottom side on

I think we do that for performance/rendering, but I think switching to regular borders for this case is fine. 🤷‍♀️

mreishus committed:
Fix issue when hovering over a selected segment

mreishus committed:
Box shadow -> Combined border

mreishus committed:
Less spacing between theme showcase search bar and tabs

mreishus commented:
This should be good for another review push now.

sixhours commented on client/my-sites/themes/theme-showcase.scss:
```diff
@@ -54,7 +54,9 @@
 .section-nav.themes__section-nav {
-	margin-top: 1px;
+	box-shadow: none;
+	border: 1px solid rgba( var( --color-neutral-10-rgb ), 0.5 );
```
Let's use our color scheme variable for the color here rather than a transparency: `var( --color-border-subtle )`

sixhours commented on client/my-sites/themes/themes-magic-search-card/style.scss:
```diff
@@ -9,8 +9,8 @@
 	background: white;
-	box-shadow: 0 0 0 1px rgba( var( --color-neutral-10-rgb ), 0.5 ),
-		0 1px 2px var( --color-neutral-0 );
+	border: 1px solid rgba( var( --color-neutral-10-rgb ), 0.5 );
```
Same note for the border color.

sixhours commented on client/my-sites/themes/themes-magic-search-card/style.scss:
```diff
@@ -158,31 +178,38 @@
 
 	text-transform: uppercase;
-	color: var( --color-neutral-70 );
+	color: var( --color-neutral-50 );
```
`var( --color-text-subtle )` would probably work here.

sixhours commented on client/my-sites/themes/themes-magic-search-card/style.scss:
```diff
@@ -158,31 +178,38 @@
 }
 
 .themes-magic-search-card__welcome-taxonomy {
+	border: 1px solid var( --color-neutral-5 );
```
`var( --color-border-subtle )` for this.

kwight commented:
Yay, this is working great! All of the elements are where they should be, for both mobile and desktop, and the experience is more sensical. 🎉

There's a number of polish issues remaining, but I'd do those as followups to keep our speed up (the more we try to get into this all-in-one PR, the longer everything takes).

<img width="578" alt="Screen Shot 2021-07-14 at 12 46 26 PM" src="https://user-images.githubusercontent.com/349751/125685725-e400ca36-a979-4d99-b2e9-38de82d57af3.png">

* Tops and bottoms of the search and segmented components aren't aligned. :heavy_check_mark: 
* The right sides of the segmented section and the current theme card button above it aren't aligned. :heavy_check_mark: 
* The icons of the filter popover are sitting visually low.
  * Matt: Possibly fixed 90b359df30, Might need someone else to try a more elegant solution though
* The space between the search and segmented components doesn't match the card's padding (it's too big). :heavy_check_mark: 
* I think the background color of the popover "search by" area should be white (if not, then the arrow above it should match instead of being white). :heavy_check_mark:

sixhours commented on client/my-sites/themes/themes-magic-search-card/style.scss:
```diff
@@ -158,31 +178,38 @@
+	border: 1px solid var( --color-neutral-5 );
+	border-left: 0;
+	&:first-child {
+		border-left: 1px solid var( --color-neutral-5 );
```
Same for the border color here.

sixhours commented on client/my-sites/themes/themes-magic-search-card/index.jsx:
```diff
@@ -307,9 +315,54 @@ class ThemesMagicSearchCard extends React.Component {
+					<div>
+						<Button
+							onClick={ this.togglePopover }
+							className="components-button themes-magic-search-card__advanced-toggle"
```
Do we need `components-button` here? Shouldn't it be part of the component's classNames list already?

mreishus replied:
It does seem to be needed, here's what it looks like with it removed:
![2021-07-14_15-34](https://user-images.githubusercontent.com/937354/125689290-557b1484-158d-4b5e-84e1-12e9bb5498d1.png)
Not exactly sure what's going on, @blackjackkent might have more insight

blackjackkent replied:
Yeah you can have a button without the component styling (as a link for instance) and it doesn't seem to be applied by default. I'll double check this tomorrow though.

kwight commented:
Mobile in different languages still look decent too.

<img width="360" alt="Screen Shot 2021-07-14 at 1 01 05 PM" src="https://user-images.githubusercontent.com/349751/125686809-7f316a03-795e-4297-bf86-2d2482354e8d.png">

kwight commented:
@ianstewart @onuro Is there anything you consider a blocker here as-is? Keep in mind, that we'll move much quicker as soon as this big main piece is deployed.

sixhours reviewed (commented):
This is looking really good! I found one usability gotcha that I'd consider a blocker, and a couple things that aren't.

I think we need to find a way to handle the filters when the list of suggestions is longer than the screen height. They get cut off and there's no way to scroll them, so you can't access anything beyond the height of the viewport:

**This PR:**

I'm on a 15" laptop screen FWIW and things get cut off around "Photoblogging"

<img width="1192" alt="Screen Shot 2021-07-14 at 3 35 51 PM" src="https://user-images.githubusercontent.com/2124984/125682680-fb9d58f8-e47d-4ea8-85de-d718f921a5fe.png">

**Production**

The rest of the list beyond Photoblogging:

<img width="1354" alt="Screen Shot 2021-07-14 at 3 36 33 PM" src="https://user-images.githubusercontent.com/2124984/125682703-2aeafbd8-b056-45c7-832a-55ad522393bf.png">

To reproduce:

* Open the filters, select Subject or Feature, both of which have 39 entries
* Click "Show all" when the search prompt opens
* You can't see the full list of subjects or features unless you're on a big, big screen

This is related to #54358 where I suggested we let the list scroll. But maybe for now we just remove the ability to "Show all" and only show a small number of subjects/features until the user has started to type?

Related, should we extend the number of column options shown to 6? It seems a little silly to have to expand the list to get a single extra result. Obviously not a blocker.

<img width="1147" alt="Screen Shot 2021-07-14 at 3 44 48 PM" src="https://user-images.githubusercontent.com/2124984/125683330-fcd16279-1eb5-448c-aa28-0c9615e1a170.png">

I also found the focus state of the search box odd, now that the search box is its own box. @onuro did you have thoughts on the updated focus state? Obviously not a blocker, either.

<img width="1364" alt="Screen Shot 2021-07-14 at 3 28 46 PM" src="https://user-images.githubusercontent.com/2124984/125684732-3a4de7bc-46e1-4eba-9947-ee1d4dfd17cf.png">

Switching between filters without making a selection populates the search with a bunch of `subject: filter: columns:` type text, which looks a little weird, especially if you're just browsing the available features:

<img width="1140" alt="Screen Shot 2021-07-14 at 3 58 54 PM" src="https://user-images.githubusercontent.com/2124984/125685458-b999b4ed-5ab8-448b-babb-587ece45d8f7.png">

Would it make more sense to clear the search after selecting a filter or subject? Or only put the prefix in when someone makes a selection? That would make it harder to search on multiple parameters...

mreishus committed:
Use border-subtle and text-subtle colors

mreishus commented:
* Updated the colors in the line by line suggestions
* Simple scrolling PR for keyed suggestions: https://github.com/Automattic/wp-calypso/pull/54567
* The focus thing is definitely weird. It's not a ':focus' css selector, it's driven off a custom `.has-highlight` class that is added based on one of the component's state variables

mreishus committed:
Fix pixel alignment error of segmented items and search bar

mreishus committed:
Remove welcome-header specific background color

mreishus committed:
Try to align text/svgs better in magic search welcome

mreishus committed:
Fix horizontal spacing

mreishus committed:
Align current theme card button info card with segmented control in magic search

sixhours reviewed (approved):
IMO the only blocker is fixed with #54567. LGTM!

mreishus committed:
Add simple scrolling bar to keyed suggestions (#54567)

ianstewart commented:
> There’s a number of polish issues remaining, but I’d do those as followups to keep our speed up (the more we try to get into this all-in-one PR, the longer everything takes).
— @kwight 

Agreed.

> @ianstewart @onuro Is there anything you consider a blocker here as-is? Keep in mind, that we'll move much quicker as soon as this big main piece is deployed.

LGTM based on agreement with that note above!
